bsdlib/doc/pdn_extension.rst

```diff
@@ -0,0 +1,174 @@
+
+
+Checking if a PDN is active
+***************************
+
+.. code-block:: c
+
+   nrf_pdn_state_t state;
+   nrf_getsockopt(pdn_fd, NRF_SOL_PDN, NRF_SO_PDN_STATE, &state, sizeof(nrf_pdn_state_t));
+
+If a data socket is bound to an APN/PDN that is closed, the socket fails with an ``NRF_ENETDOWN`` error.
+
+The example code shows how to check whether a PDN socket is still active.
+However, if there is a long time between transmissions, the PDN socket might appear to still be active, even though the network has deactivated the PDN.
+Only checking the state of the PDN by querying the socket option NRF_SO_PDN_STATE is therefore not reliable.
+Instead, send a transmission on the data socket to trigger the link to become active again.
+The network will then report if the PDN has been deactivated.
+In that case, you must create another PDN socket.
+
+.. note::
+   Setting the socket option NRF_SO_PDN_STATE with ``nrf_setsockopt`` to manually control activation/deactivation not supported.
+
+
+Retrieving the context ID (CID)
+*******************************
+
+.. code-block:: c
+
+   nrf_pdn_context_id_t cid;
+   nrf_getsockopt(pdn_fd, NRF_SOL_PDN, NRF_SO_PDN_CONTEXT_ID, &cid, sizeof(nrf_pdn_context_id_t));
+
+In some occasions, it is useful to know the CID that is assigned to a PDN socket.
+To retrieve the CID, get the socket option NRF_SO_PDN_CONTEXT_ID with ``nrf_getsockopt``.
+
+NRF_SO_PDN_CONTEXT_ID is a read-only socket option.
+The CID is an internal allocation and cannot be configured by the user.
+
+
+Sending a DNS query
+*******************
+
+.. code-block:: c
+
+   struct nrf_addrinfo hints = {
+     .ai_family = NRF_AF_INET,
+     .ai_socktype = NRF_SOCK_STREAM
+   }
+
+   struct nrf_addrinfo apn_hints;
+
+   apn_hints.ai_family = NRF_AF_LTE;
+   apn_hints.ai_socktype = NRF_SOCK_MGMT;
+   apn_hints.ai_protocol = NRF_PROTO_PDN;
+   apn_hints.ai_canonname = "custom.apn";
+
+   hints.ai_next = &apn_hints;
+
+   nrf_getaddrinfo("example.com", NULL, &hints, &result);
+
+The example code shows how to specify an APN to use for a DNS query.
+In this example, the DNS hints are extended with an extra linked-list hint that specifies the APN to use for the query.
+
+Like for data sockets, the APN name must be set up and activated before the DNS query.
+Otherwise, the query will fail.
+
+Waiting for an IPv6 connection
+******************************
+
+.. code-block:: c
+
+   at_fd = nrf_socket(NRF_AF_LTE, 0, NRF_NPROTO_AT);
+   nrf_send(at_fd, "AT+CGEREP=1", strlen("AT+CGEREP=1"));
+   # -> AT reponse "OK"
+
+   # Set up PDN socket using the APN (PDN initialize and connect)
+
+   nrf_getsockopt(pdn_fd, NRF_SOL_PDN, NRF_SO_PDN_CONTEXT_ID, &context_id, sizeof(uint8_t));
+   while not (context_id in ":CGEV IPV6 <cid>");
+   nrf_close(at_fd);
+
+When requesting an IPv6 connection (which requires PDN type IPV6 or IPV4V6), the IPv6 address might not be correct even if the request is successful.
```

Comment:
This paragraph could precede the code snippet and the one below immediately follow it